### PR TITLE
End-of-Life behaviour (EXPOSUREAPP-4641)

### PR DESCRIPTION
**Pull request by jurajkusnier**
### End of Life behaviour
- Stop risk calculation when the user got a positive test result
- Reenable risk calculation & delete test when the user switches back to "normal" mode
- Don't stop key downloading & exposure check

Fixing also EXPOSUREAPP-3981: Reminder Open App although End-Of-Life reached

### CODE REVIEW

**Review by SamuraiKek — DISMISSED** [review]
Looks good to me. Good catch on the spelling mistakes and unnecessary true assertions.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/ui/homecards/SubmissionState.kt [hunk 1]
```diff
@@ -10,4 +15,9 @@ object TestNegative : SubmissionState()
 object TestError : SubmissionState()
 object TestInvalid : SubmissionState()
 object TestPending : SubmissionState()
-object SubmissionDone : SubmissionState()
+data class SubmissionDone(val testRegisteredOn: Date?) : SubmissionState() {
```
Comment (chiljamgossow) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/ui/homecards/SubmissionState.kt:18]: Does the date need to be nullable?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragmentViewModel.kt [hunk 2]
```diff
@@ -221,14 +224,22 @@ class HomeFragmentViewModel @AssistedInject constructor(
...
-                TestPositive, SubmissionDone -> {
+                TestPositive, is SubmissionDone -> {
                     // Don't show risk card
                 }
                 else -> add(tracingItem)
             }
 
             add(submissionItem)
 
+            if (submissionState is SubmissionDone) {
```
Comment (chiljamgossow) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragmentViewModel.kt:235]: Shouldn't we show the card once the user has seen the positive test result? The ticket says: Once the user got a positive test result, the risk calculation shall be stopped.
Submission done comes later and may never be reached.

**Review by chiljamgossow — CHANGES_REQUESTED** [review]
When the submission is done, I can see the new card which tells me that risk calculation is off. At the same time, the app bar says it is still running, and I can turn it on and off in the setting.
The popup that comes up when pressing 'einschalten' only talks about removing a test, which is confusing.
When I add a new test, the result screen is broken.

**Comment by chiljamgossow** [comment]
The app flow seems to be broken after removing the test with the new card.
It works fine if I use reset in the settings.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragmentViewModel.kt (hunk 2 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragmentViewModel.kt:235]: No, the ticket says that we should inform users after they share their keys

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/ui/homecards/SubmissionState.kt (hunk 1 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/ui/homecards/SubmissionState.kt]: no

**Comment by jurajkusnier** [comment]
> The app flow seems to be broken after removing the test with the new card.

What do you exactly mean?

**Comment by chiljamgossow** [comment]
> > The app flow seems to be broken after removing the test with the new card.
> 
> What do you exactly mean?

Reenable via the new card and then scan a new positive test. The app is not working correctly any more.

**Comment by SamuraiKek** [comment]
> When the submission is done, I can see the new card which tells me that risk calculation is off. At the same time, the app bar says it is still running, and I can turn it on and off in the setting.
I agree, the app bar should also be deactivated. I'm wondering what should happen if we tried to activate the exposure logging from the "Exposure logging" screen. Should it have the same behaviour as pressing the "Einschalten" button on the new card or should we completely disable activation from this screen and only do it from the new card.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragmentViewModel.kt (hunk 2 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragmentViewModel.kt:235]: Yes, you are right

**Comment by jurajkusnier** [comment]
> > > The app flow seems to be broken after removing the test with the new card.
> > 
> > 
> > What do you exactly mean?
> 
> Reenable via the new card and then scan a new positive test. The app is not working correctly any more.

Could you try with the latest changes? The only difference I noticed when I was testing 2 positive tests in a row, is that the ENF client doesn't ask you for consent to share exposure keys because it caches your previous consent for 24h. That's not an issue but expected behavior.

**Review by chiljamgossow — DISMISSED** [review]
I tested again and the app seems to work fine now after enabling exposure logging again (before the test result screen was incomplete)

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragmentViewModel.kt [hunk 3]
```diff
@@ -224,14 +227,22 @@ class HomeFragmentViewModel @AssistedInject constructor(
     ) { tracingItem, submissionItem, submissionState, statsData ->
         mutableListOf<HomeItem>().apply {
             when (submissionState) {
-                TestPositive, SubmissionDone -> {
+                TestPositive, is SubmissionDone -> {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragmentViewModel.kt:230]: What does this change do?

File: Corona-Warn-App/src/main/res/layout/home_reenable_risk_card_layout.xml [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        <TextView
+            android:id="@+id/reenable_risk_card_description"
+            style="@style/body2"
+            android:layout_width="0dp"
+            android:layout_height="wrap_content"
+            android:layout_marginTop="@dimen/spacing_medium"
+            android:layout_marginEnd="@dimen/spacing_small"
+            android:text="@string/reenable_risk_card_description_text"
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintTop_toBottomOf="@id/reenable_risk_card_divider" />
```
Comment (d4rken) [inline Corona-Warn-App/src/main/res/layout/home_reenable_risk_card_layout.xml:92]: Top and bottom text have the same style (`body2`), while in the mockups the bottom text has a larger font size. Intentional change?

Screenshot from 2021-02-08 15-40-21

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragmentViewModel.kt (hunk 3 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragmentViewModel.kt:230]: `SubmissionDone` is now data class not an object, so we need to change this syntax.

File: Corona-Warn-App/src/main/res/layout/home_reenable_risk_card_layout.xml (hunk 4 above)
Comment (SamuraiKek) [reply Corona-Warn-App/src/main/res/layout/home_reenable_risk_card_layout.xml:92]: Didn't catch that. It's indeed a bit bigger in Figma.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragmentViewModel.kt (hunk 3 above)
Comment (SamuraiKek) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragmentViewModel.kt:230]: I've changed SubmissionDone from an object to a data class so we can no longer use it without explicitly putting the is operator before it.

File: Corona-Warn-App/src/main/res/layout/home_reenable_risk_card_layout.xml (hunk 4 above)
Comment (SamuraiKek) [reply Corona-Warn-App/src/main/res/layout/home_reenable_risk_card_layout.xml]: I've increased the font size to match Figma.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/SubmissionRepository.kt [hunk 5]
```diff
@@ -182,6 +182,7 @@ class SubmissionRepository @Inject constructor(
         LocalData.initialTestResultReceivedTimestamp(0L)
         LocalData.isAllowedToSubmitDiagnosisKeys(false)
         LocalData.isTestResultAvailableNotificationSent(false)
+        LocalData.numberOfSuccessfulSubmissions(0)
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/SubmissionRepository.kt:185]: There might be an issue with the state still.

* During positive test, key download should keep running for plausible deniability?
* After removing the positive test key download is only enabled after restarting the app

### Before first test
```bash
adb shell am broadcast -a "androidx.work.diagnostics.REQUEST_DIAGNOSTICS" -p de.rki.coronawarnapp.test

2021-02-08 19:37:17.065 21765-22027/de.rki.coronawarnapp.test I/WM-DiagnosticsWrkr: Recently completed work:
2021-02-08 19:37:17.070 21765-22027/de.rki.coronawarnapp.test I/WM-DiagnosticsWrkr:  Id 	 Class Name	 Job Id	 State	 Unique Name	 Tags	
    e79c3dcb-2c78-40ab-9301-2bce816fbf60	 de.rki.coronawarnapp.worker.DiagnosisKeyRetrievalOneTimeWorker	 null	 SUCCEEDED	 DiagnosisKeyBackgroundOneTimeWork	 DIAGNOSIS_KEY_ONE_TIME_WORKER,de.rki.coronawarnapp.worker.DiagnosisKeyRetrievalOneTimeWorker	
    22838970-130e-4cb8-961f-f02d7cea3ea1	 de.rki.coronawarnapp.nearby.ExposureStateUpdateWorker	 null	 SUCCEEDED	 	 de.rki.coronawarnapp.nearby.ExposureStateUpdateWorker	
2021-02-08 19:37:17.070 21765-22027/de.rki.coronawarnapp.test I/WM-DiagnosticsWrkr: Running work:
2021-02-08 19:37:17.072 21765-22027/de.rki.coronawarnapp.test I/WM-DiagnosticsWrkr:  Id 	 Class Name	 Job Id	 State	 Unique Name	 Tags	
    24a5b31f-37ee-4836-a9ae-f647f1543af2	 androidx.work.impl.workers.DiagnosticsWorker	 9	 RUNNING	 	 androidx.work.impl.workers.DiagnosticsWorker	
2021-02-08 19:37:17.072 21765-22027/de.rki.coronawarnapp.test I/WM-DiagnosticsWrkr: Enqueued work:
2021-02-08 19:37:17.080 21765-22027/de.rki.coronawarnapp.test I/WM-DiagnosticsWrkr:  Id 	 Class Name	 Job Id	 State	 Unique Name	 Tags	
    0693b399-bf19-43a2-aea3-1e6006f0734f	 de.rki.coronawarnapp.worker.DiagnosisKeyRetrievalPeriodicWorker	 5	 ENQUEUED	 DiagnosisKeyBackgroundPeriodicWork	 DIAGNOSIS_KEY_PERIODIC_WORKER,de.rki.coronawarnapp.worker.DiagnosisKeyRetrievalPeriodicWorker	
    eae464f2-0163-4f3c-ba6a-8f371c43ab8c	 de.rki.coronawarnapp.deadman.DeadmanNotificationOneTimeWorker	 6	 ENQUEUED	 DeadmanNotificationOneTimeWork	 de.rki.coronawarnapp.deadman.DeadmanNotificationOneTimeWorker	
    e3ab1f79-8b33-4883-8287-e65c9663dcf1	 de.rki.coronawarnapp.deadman.DeadmanNotificationPeriodicWorker	 7	 ENQUEUED	 DeadmanNotificationPeriodicWork	 de.rki.coronawarnapp.deadman.DeadmanNotificationPeriodicWorker	
    ee7383df-587a-4bce-91ad-8cc3ea9be327	 de.rki.coronawarnapp.contactdiary.retention.ContactDiaryRetentionWorker	 8	 ENQUEUED	 ContactDiaryPeriodicWork	 de.rki.coronawarnapp.contactdiary.retention.ContactDiaryRetentionWorker	

```

### Positive test and submission
```bash
adb shell am broadcast -a "androidx.work.diagnostics.REQUEST_DIAGNOSTICS" -p de.rki.coronawarnapp.test

2021-02-08 19:38:58.558 21765-22170/de.rki.coronawarnapp.test I/WM-DiagnosticsWrkr: Recently completed work:
2021-02-08 19:38:58.573 21765-22170/de.rki.coronawarnapp.test I/WM-DiagnosticsWrkr:  Id 	 Class Name	 Job Id	 State	 Unique Name	 Tags	
    ec4eb9da-0466-43b5-a8c4-a024f63df9de	 de.rki.coronawarnapp.submission.auto.SubmissionWorker	 null	 CANCELLED	 AutoSubmissionWorker	 AutoSubmissionWorker,de.rki.coronawarnapp.submission.auto.SubmissionWorker	
    24a5b31f-37ee-4836-a9ae-f647f1543af2	 androidx.work.impl.workers.DiagnosticsWorker	 null	 SUCCEEDED	 	 androidx.work.impl.workers.DiagnosticsWorker	
    e3ab1f79-8b33-4883-8287-e65c9663dcf1	 de.rki.coronawarnapp.deadman.DeadmanNotificationPeriodicWorker	 null	 CANCELLED	 DeadmanNotificationPeriodicWork	 de.rki.coronawarnapp.deadman.DeadmanNotificationPeriodicWorker	
    eae464f2-0163-4f3c-ba6a-8f371c43ab8c	 de.rki.coronawarnapp.deadman.DeadmanNotificationOneTimeWorker	 null	 CANCELLED	 DeadmanNotificationOneTimeWork	 de.rki.coronawarnapp.deadman.DeadmanNotificationOneTimeWorker	
    0693b399-bf19-43a2-aea3-1e6006f0734f	 de.rki.coronawarnapp.worker.DiagnosisKeyRetrievalPeriodicWorker	 null	 CANCELLED	 DiagnosisKeyBackgroundPeriodicWork	 DIAGNOSIS_KEY_PERIODIC_WORKER,de.rki.coronawarnapp.worker.DiagnosisKeyRetrievalPeriodicWorker	
    e79c3dcb-2c78-40ab-9301-2bce816fbf60	 de.rki.coronawarnapp.worker.DiagnosisKeyRetrievalOneTimeWorker	 null	 SUCCEEDED	 DiagnosisKeyBackgroundOneTimeWork	 DIAGNOSIS_KEY_ONE_TIME_WORKER,de.rki.coronawarnapp.worker.DiagnosisKeyRetrievalOneTimeWorker	
    22838970-130e-4cb8-961f-f02d7cea3ea1	 de.rki.coronawarnapp.nearby.ExposureStateUpdateWorker	 null	 SUCCEEDED	 	 de.rki.coronawarnapp.nearby.ExposureStateUpdateWorker	
2021-02-08 19:38:58.573 21765-22170/de.rki.coronawarnapp.test I/WM-DiagnosticsWrkr: Running work:
2021-02-08 19:38:58.575 21765-22170/de.rki.coronawarnapp.test I/WM-DiagnosticsWrkr:  Id 	 Class Name	 Job Id	 State	 Unique Name	 Tags	
    4af8a51a-a833-41ae-9b1e-9874e67399c7	 androidx.work.impl.workers.DiagnosticsWorker	 15	 RUNNING	 	 androidx.work.impl.workers.DiagnosticsWorker	
2021-02-08 19:38:58.575 21765-22170/de.rki.coronawarnapp.test I/WM-DiagnosticsWrkr: Enqueued work:
2021-02-08 19:38:58.578 21765-22170/de.rki.coronawarnapp.test I/WM-DiagnosticsWrkr:  Id 	 Class Name	 Job Id	 State	 Unique Name	 Tags	
    01cc12e3-64f7-4b54-96a7-06e58ae52e8b	 de.rki.coronawarnapp.contactdiary.retention.ContactDiaryRetentionWorker	 12	 ENQUEUED	 ContactDiaryPeriodicWork	 de.rki.coronawarnapp.contactdiary.retention.ContactDiaryRetentionWorker	
```

### Test removed and tracing re-enabled
```bash
adb shell am broadcast -a "androidx.work.diagnostics.REQUEST_DIAGNOSTICS" -p de.rki.coronawarnapp.test

2021-02-08 19:40:16.510 22200-22331/de.rki.coronawarnapp.test I/WM-DiagnosticsWrkr: Recently completed work:
2021-02-08 19:40:16.536 22200-22331/de.rki.coronawarnapp.test I/WM-DiagnosticsWrkr:  Id 	 Class Name	 Job Id	 State	 Unique Name	 Tags	
    40785e3c-aae7-4605-adc6-0cd4f571e915	 androidx.work.impl.workers.DiagnosticsWorker	 null	 SUCCEEDED	 	 androidx.work.impl.workers.DiagnosticsWorker	
    8892a4df-0edb-45c2-b040-be70e1b45a93	 androidx.work.impl.workers.DiagnosticsWorker	 16	 SUCCEEDED	 	 androidx.work.impl.workers.DiagnosticsWorker	
    4af8a51a-a833-41ae-9b1e-9874e67399c7	 androidx.work.impl.workers.DiagnosticsWorker	 null	 SUCCEEDED	 	 androidx.work.impl.workers.DiagnosticsWorker	
    ec4eb9da-0466-43b5-a8c4-a024f63df9de	 de.rki.coronawarnapp.submission.auto.SubmissionWorker	 null	 CANCELLED	 AutoSubmissionWorker	 AutoSubmissionWorker,de.rki.coronawarnapp.submission.auto.SubmissionWorker	
    24a5b31f-37ee-4836-a9ae-f647f1543af2	 androidx.work.impl.workers.DiagnosticsWorker	 null	 SUCCEEDED	 	 androidx.work.impl.workers.DiagnosticsWorker	
    eae464f2-0163-4f3c-ba6a-8f371c43ab8c	 de.rki.coronawarnapp.deadman.DeadmanNotificationOneTimeWorker	 null	 CANCELLED	 DeadmanNotificationOneTimeWork	 de.rki.coronawarnapp.deadman.DeadmanNotificationOneTimeWorker	
    0693b399-bf19-43a2-aea3-1e6006f0734f	 de.rki.coronawarnapp.worker.DiagnosisKeyRetrievalPeriodicWorker	 null	 CANCELLED	 DiagnosisKeyBackgroundPeriodicWork	 DIAGNOSIS_KEY_PERIODIC_WORKER,de.rki.coronawarnapp.worker.DiagnosisKeyRetrievalPeriodicWorker	
    e79c3dcb-2c78-40ab-9301-2bce816fbf60	 de.rki.coronawarnapp.worker.DiagnosisKeyRetrievalOneTimeWorker	 null	 SUCCEEDED	 DiagnosisKeyBackgroundOneTimeWork	 DIAGNOSIS_KEY_ONE_TIME_WORKER,de.rki.coronawarnapp.worker.DiagnosisKeyRetrievalOneTimeWorker	
    22838970-130e-4cb8-961f-f02d7cea3ea1	 de.rki.coronawarnapp.nearby.ExposureStateUpdateWorker	 null	 SUCCEEDED	 	 de.rki.coronawarnapp.nearby.ExposureStateUpdateWorker	
2021-02-08 19:40:16.536 22200-22331/de.rki.coronawarnapp.test I/WM-DiagnosticsWrkr: Running work:
2021-02-08 19:40:16.538 22200-22331/de.rki.coronawarnapp.test I/WM-DiagnosticsWrkr:  Id 	 Class Name	 Job Id	 State	 Unique Name	 Tags	
    fdc2204f-6c4d-45f5-94b8-7ae412fd2a54	 androidx.work.impl.workers.DiagnosticsWorker	 21	 RUNNING	 	 androidx.work.impl.workers.DiagnosticsWorker	
2021-02-08 19:40:16.538 22200-22331/de.rki.coronawarnapp.test I/WM-DiagnosticsWrkr: Enqueued work:
2021-02-08 19:40:16.543 22200-22331/de.rki.coronawarnapp.test I/WM-DiagnosticsWrkr:  Id 	 Class Name	 Job Id	 State	 Unique Name	 Tags	
    d97ceb3a-bd2c-49b2-8222-b60f64bffc69	 de.rki.coronawarnapp.contactdiary.retention.ContactDiaryRetentionWorker	 18	 ENQUEUED	 ContactDiaryPeriodicWork	 de.rki.coronawarnapp.contactdiary.retention.ContactDiaryRetentionWorker	
    284e1d7b-ca04-412e-bbee-4b0a939b500c	 de.rki.coronawarnapp.deadman.DeadmanNotificationPeriodicWorker	 20	 ENQUEUED	 DeadmanNotificationPeriodicWork	 de.rki.coronawarnapp.deadman.DeadmanNotificationPeriodicWorker	
```

### Test removed, tracing re-enabled and app restarted
```bash
adb shell am broadcast -a "androidx.work.diagnostics.REQUEST_DIAGNOSTICS" -p de.rki.coronawarnapp.test

2021-02-08 19:41:16.755 22379-22521/de.rki.coronawarnapp.test I/WM-DiagnosticsWrkr: Recently completed work:
2021-02-08 19:41:16.784 22379-22521/de.rki.coronawarnapp.test I/WM-DiagnosticsWrkr:  Id 	 Class Name	 Job Id	 State	 Unique Name	 Tags	
    f5f3036d-2c32-4d59-9511-fb2038b5883c	 androidx.work.impl.workers.DiagnosticsWorker	 null	 SUCCEEDED	 	 androidx.work.impl.workers.DiagnosticsWorker	
    fdc2204f-6c4d-45f5-94b8-7ae412fd2a54	 androidx.work.impl.workers.DiagnosticsWorker	 null	 SUCCEEDED	 	 androidx.work.impl.workers.DiagnosticsWorker	
    40785e3c-aae7-4605-adc6-0cd4f571e915	 androidx.work.impl.workers.DiagnosticsWorker	 null	 SUCCEEDED	 	 androidx.work.impl.workers.DiagnosticsWorker	
    8892a4df-0edb-45c2-b040-be70e1b45a93	 androidx.work.impl.workers.DiagnosticsWorker	 16	 SUCCEEDED	 	 androidx.work.impl.workers.DiagnosticsWorker	
    4af8a51a-a833-41ae-9b1e-9874e67399c7	 androidx.work.impl.workers.DiagnosticsWorker	 null	 SUCCEEDED	 	 androidx.work.impl.workers.DiagnosticsWorker	
    ec4eb9da-0466-43b5-a8c4-a024f63df9de	 de.rki.coronawarnapp.submission.auto.SubmissionWorker	 null	 CANCELLED	 AutoSubmissionWorker	 AutoSubmissionWorker,de.rki.coronawarnapp.submission.auto.SubmissionWorker	
    24a5b31f-37ee-4836-a9ae-f647f1543af2	 androidx.work.impl.workers.DiagnosticsWorker	 null	 SUCCEEDED	 	 androidx.work.impl.workers.DiagnosticsWorker	
    eae464f2-0163-4f3c-ba6a-8f371c43ab8c	 de.rki.coronawarnapp.deadman.DeadmanNotificationOneTimeWorker	 null	 CANCELLED	 DeadmanNotificationOneTimeWork	 de.rki.coronawarnapp.deadman.DeadmanNotificationOneTimeWorker	
    e79c3dcb-2c78-40ab-9301-2bce816fbf60	 de.rki.coronawarnapp.worker.DiagnosisKeyRetrievalOneTimeWorker	 null	 SUCCEEDED	 DiagnosisKeyBackgroundOneTimeWork	 DIAGNOSIS_KEY_ONE_TIME_WORKER,de.rki.coronawarnapp.worker.DiagnosisKeyRetrievalOneTimeWorker	
    22838970-130e-4cb8-961f-f02d7cea3ea1	 de.rki.coronawarnapp.nearby.ExposureStateUpdateWorker	 null	 SUCCEEDED	 	 de.rki.coronawarnapp.nearby.ExposureStateUpdateWorker	
2021-02-08 19:41:16.784 22379-22521/de.rki.coronawarnapp.test I/WM-DiagnosticsWrkr: Running work:
2021-02-08 19:41:16.786 22379-22521/de.rki.coronawarnapp.test I/WM-DiagnosticsWrkr:  Id 	 Class Name	 Job Id	 State	 Unique Name	 Tags	
    5ed01f56-e6f9-4845-8d02-42d815460523	 androidx.work.impl.workers.DiagnosticsWorker	 27	 RUNNING	 	 androidx.work.impl.workers.DiagnosticsWorker	
2021-02-08 19:41:16.786 22379-22521/de.rki.coronawarnapp.test I/WM-DiagnosticsWrkr: Enqueued work:
2021-02-08 19:41:16.793 22379-22521/de.rki.coronawarnapp.test I/WM-DiagnosticsWrkr:  Id 	 Class Name	 Job Id	 State	 Unique Name	 Tags	
    284e1d7b-ca04-412e-bbee-4b0a939b500c	 de.rki.coronawarnapp.deadman.DeadmanNotificationPeriodicWorker	 20	 ENQUEUED	 DeadmanNotificationPeriodicWork	 de.rki.coronawarnapp.deadman.DeadmanNotificationPeriodicWorker	
    23d6eb6b-4d9f-49c4-a245-6b43b613399c	 de.rki.coronawarnapp.worker.DiagnosisKeyRetrievalPeriodicWorker	 24	 ENQUEUED	 DiagnosisKeyBackgroundPeriodicWork	 DIAGNOSIS_KEY_PERIODIC_WORKER,de.rki.coronawarnapp.worker.DiagnosisKeyRetrievalPeriodicWorker	
    3e2276a6-a220-4788-9294-cad09c7e286f	 de.rki.coronawarnapp.contactdiary.retention.ContactDiaryRetentionWorker	 26	 ENQUEUED	 ContactDiaryPeriodicWork	 de.rki.coronawarnapp.contactdiary.retention.ContactDiaryRetentionWorker	
```

**Review by d4rken — CHANGES_REQUESTED** [review]
Looks good when testing on device, but looking at the workers there might still be an issue with the scheduling, could you review my logs and the specs?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/SubmissionRepository.kt (hunk 5 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/SubmissionRepository.kt:185]: Yes, the key download should be running all the time. It's fixed in the future commit.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/task/SubmissionTask.kt [hunk 6]
```diff
@@ -154,6 +154,7 @@ class SubmissionTask @Inject constructor(
         Timber.tag(TAG).d("setSubmissionFinished()")
         BackgroundWorkScheduler.stopWorkScheduler()
         LocalData.numberOfSuccessfulSubmissions(1)
+        BackgroundWorkScheduler.startWorkScheduler()
```
[the remark this answers is not in the document]
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/task/SubmissionTask.kt:157]: We need to keep the diagnostic job running but we should kill other workers like the test result retriever.
This is one way how to implement it, another way would be stoping only a filtered list of workers. The first implementation seems cleaner to me.

**Review by d4rken — APPROVED** [review]
nice! lgtm

**Review by chiljamgossow — APPROVED** [review]

**Review by janetback — APPROVED** [review]
UA approved

**Pull request merged**